### ISSUE TITLE
non_negative_derivative handles one piece of data unexpectedly (huge spikes to 400 trillion) 

### ISSUE THREAD

**Issue by daviesalex** [influxdata/influxdb#9252]
### Bug report

__System info:__ 1.4.2 on Linux

__Steps to reproduce:__

Create a series with incrementing counter, but have a single item significantly lower, e.g.:
```
> SELECT max(tx_bytes) FROM "netif" WHERE "interface" = 'enp129s0f1' and "host" = 'fpia-glacial02' and time > 1513426480664ms and time < 1513426551084ms GROUP BY time(10s),interface, "host" fill(none);
```
``` name: netif
tags: host=fpia-glacial02, interface=enp129s0f1
time                max
----                ---
1513426490000000000 838713498922297
1513426500000000000 838713498922641
1513426510000000000 280854869492039
1513426520000000000 838713498922899
1513426530000000000 838713498923243
1513426540000000000 838713498923333
1513426550000000000 838713498923423
```

2. Ask for the non_negative_derivative() of this series

```
> SELECT non_negative_derivative(max(tx_bytes), 1s) * 8 FROM "netif" WHERE "interface" = 'enp129s0f1' and "host" = 'fpia-glacial02' and time > 1513426480664ms and time < 1513426551084ms GROUP BY time(10s),interface, "host" fill(previous);
```
``` name: netif
tags: host=fpia-glacial02, interface=enp129s0f1
time                non_negative_derivative
----                -----------------------
1513426500000000000 275.2
1513426520000000000 446286903544688
1513426530000000000 275.2
1513426540000000000 72
1513426550000000000 72
```

3. Observe that the second value is clearly crazy (~400 trillion)

__Expected behavior:__
The second derivative entry should not exist at all - it should be ignored as clearly erroneous (as the negative derivative is). 

It would not be unexpected (but not desirable) if it is some utterly enormous number (~infinity)

It is very unclear where this current number comes from; presumably some function of the maximum value of a data structure and the precise way its being calculated. 

__Actual behavior:__ [What actually happened]
Some number around 400 trillion is returned (not always the same number)

__Additional info:__ 
This is often hit with counters 'rolling over' to 0 and then going back to some positive number, although in this case we hit it with obviously erroneous data. The impact is that Grafana auto-scaling does not work which is annoying. 

We are tracking back the cause of these bad counters internally, but it would really help if InfluxDB handled this (realistically common) case elegantly. In the case of 32 bit counters, often nothing can be done to prevent them rolling over.

**Comment by phemmer** [comment]
It sounds like you're asking for `non_negative_derivative()` to not emit values greater than a certain amount. What about people who actually have valid values this large? Routers can process terrabytes of data a day. A simple traffic counter would easily generate numbers this large.

**Comment by oplehto** [comment]
How about having an optional argument (or even a variant of the non_negative_derivative) to set a specific high watermark? This would avoid making this a breaking change. 

I would expect that this is a fairly common issue and therefore that having a special feature like that would be justified.

**Comment by daviesalex** [comment]
@phemmer I dont think this is about having some hard limit. When you call non_negative_derivative() on a series, but you end up with a negative derivative, what it appears to do today is skip the time period that it goes negative in the output, and then make up an arbitrary derivative for the period after that. It should handle that gracefully - either skipping both, or figuring out what to do based on the "good" periods. 

I agree that 32 bit counters can cause this (its mentioned at the top) - we have the same problem there (400 trillion appearing for a single point which renders the whole graph useless). We have almost entirely moved to equipment with 64 bit counters to fix this, but where we have them we can see it. You want exactly the same fix there - handle the rollover gracefully.

**Comment by phemmer** [comment]
> and then make up an arbitrary derivative for the period after that

Can you explain what you mean by this. I don't follow

**Comment by daviesalex** [comment]
@phemmer the query returns these derivatives:

```
1513426500000000000 275.2
<nothing is returned for time period 1513426510000000000>
1513426520000000000 446286903544688
1513426530000000000 275.2
1513426540000000000 72
```

I have no idea where 446286903544688 comes from, but its as far as I can tell just an arbitrary big number.  Its not the derivative from 0 to 838713498922899 (or 838713498922899 - 280854869492039) as you might expect, its not 0, its not the maximum value of any data type I can think of and so on.

**Comment by phemmer** [comment]
When I calculate your query by hand, I end up with the exact same value:
```
( ( 838713498922899 - 280854869492039 ) / ( ( 1513426520000000000 - 1513426510000000000 ) / 1000000000 ) ) * 8 = 446286903544688
```

**Comment by daviesalex** [comment]
@phemmer that does actually make total sense; its not coming from zero - but from that previous value (it ignores the derivative thats negative, not the point that caused it to go negative). So we go -big number then +big number. And it eats the first one because its non_negative. Thanks!

This does mean that in the case of a normal counter, this would not be a problem because the only "Big number" would be the negative, which gets eaten. What is unusual about this particular problem is that the counter is broken so its just going (massively) down for a single point and then massively up which causes this enormous value. 

We deployed a Kapacitor script yesterday to look to see if this is happening outside the specific example I found yesterday with a crazy Grafana graph.

hmm...  I wonder if there is a way to handle this in the InfluxDB side elegantly. Like an option to skip the point after a negative derivative.
